pydicom/valuerep.py

```diff
@@ -774,9 +774,9 @@ class TM(_DateTimeBase, datetime.time):
     """
     _RE_TIME = re.compile(
         r"(?P<h>^([01][0-9]|2[0-3]))"
-        r"((?P<m>([0-5][0-9]))?"
-        r"(?(5)(?P<s>([0-5][0-9]|60))?)"
-        r"(?(7)(\.(?P<ms>([0-9]{1,6})?))?))$"
+        r"((?P<m>([0-5][0-9]))"
+        r"((?P<s>([0-5][0-9]|60))"
+        r"(\.(?P<ms>([0-9]{1,6})?))?)?)?$"
```

Comment:
@mrbean-bremen I'm sorry to bring this up but could you please elaborate why this pattern works?

Looking at it I think that it's not the same as it used to be, e.g. I think that in some cases it allows for `<ms>` group to be matched even if the `<s>` one did not match. Previously this `<ms>` group matching was dependent on the success of matching `<s>` group, whereas now it's not.

Also, I wonder whether this wasn't (isn't as the problem is still present in the just released Python version) a bug in Python `re` module? The previous regex is correctly recognized by some other regexp implementation, e.g. https://regex101.com/

Reply:
> I think that in some cases it allows for <ms> group to be matched even if the `<s>` one did not match

I don't see this - as far as I understand, the <ms> group is an optional part of the expression above, while the `<s>` group is mandatory (in that scope, it is optional in the scope of the `<m>` group). Not sure if that was clear :) Anyway, if you can show me an example that would now incorrectly match, I stand corrected, of course.

> Also, I wonder whether this isn't a bug in Python re module?

Yes, it is - but I think you filed it yourself? Anywhere, it will eventually be fixed, but it is not fixed in Python 3.11.0, so we cannot remove the fix yet. In case the fix (in pydicom) is incorrect, we have to change it of course, otherwise we can just leave it as is.

Reply:
> > I think that in some cases it allows for  group to be matched even if the `<s>` one did not match
> 
> I don't see this - as far as I understand, the group is an optional part of the expression above, while the `<s>` group is mandatory (in that scope, it is optional in the scope of the `<m>` group). Not sure if that was clear :) Anyway, if you can show me an example that would now incorrectly match, I stand corrected, of course.

Ah you're right! Somehow I didn't notice the removal of the `?` on the `<s>` group...

> > Also, I wonder whether this isn't a bug in Python re module?
> 
> Yes, it is - but I think you filed it yourself? Anywhere, it will eventually be fixed, but it is not fixed in Python 3.11.0, so we cannot remove the fix yet. In case the fix (in pydicom) is incorrect, we have to change it of course, otherwise we can just leave it as is.


Yes, I have filled in a bug report once I've confirmed that there is a problem with Python's `re` module, but that was after posting the question here. Sorry for making it look so confusing :sweat_smile:

Reply:
> Yes, I have filled in a bug report once I've confirmed that there is a problem with Python's re module

Thanks for that, by the way, should have done this myself!

> Sorry for making it look so confusing

It got me to re-check the fix, which is always a good thing. Anyway, I understand that leaving the fix in is ok, regardless of the Python `re` fix, so all is well.